src/libraries/System.Security.Cryptography.Algorithms/tests/Rfc2898OneShotTests.cs

```diff
@@ -0,0 +1,321 @@
+            Assert.Equal(key1, key2);
+
+            Span<byte> destinationBuffer = new byte[length + 2];
+            Span<byte> destination = destinationBuffer.Slice(1, length);
+            Rfc2898DeriveBytes.Pbkdf2(password, salt, destination, iterations, hashAlgorithmName);
+
+            Assert.True(key1.AsSpan().SequenceEqual(destination), "key1 == destination");
+            Assert.Equal(0, destinationBuffer[^1]); // Make sure we didn't write past the destination
+            Assert.Equal(0, destinationBuffer[0]); // Make sure we didn't write before the destination
+        }
+
+        [Theory]
+        [MemberData(nameof(Pbkdf2_PasswordString_Compare_Data))]
+        public static void Pbkdf2_PasswordString_Compare(
+            string hashAlgorithm,
+            int length,
+            int iterations,
+            string password,
+            string saltHex)
+        {
+            byte[] salt = Convert.FromHexString(saltHex);
+            HashAlgorithmName hashAlgorithmName = new HashAlgorithmName(hashAlgorithm);
+            byte[] key1;
+
+            using (Rfc2898DeriveBytes instanceKdf = new Rfc2898DeriveBytes(password, salt, iterations, hashAlgorithmName))
+            {
+                key1 = instanceKdf.GetBytes(length);
+            }
+
+            // byte array allocating
+            byte[] key2 = Rfc2898DeriveBytes.Pbkdf2(password, salt, iterations, hashAlgorithmName, length);
+            Assert.Equal(key1, key2);
+
+            Span<byte> destinationBuffer = new byte[length + 2];
+            Span<byte> destination = destinationBuffer.Slice(1, length);
+            Rfc2898DeriveBytes.Pbkdf2(password, salt, destination, iterations, hashAlgorithmName);
+
+            Assert.True(key1.AsSpan().SequenceEqual(destination), "key1 == destination");
+            Assert.Equal(0, destinationBuffer[^1]); // Make sure we didn't write past the destination
+            Assert.Equal(0, destinationBuffer[0]); // Make sure we didn't write before the destination
+        }
+
+        public static IEnumerable<object[]> Pbkdf2_PasswordBytes_Compare_Data()
+        {
+            foreach (HashAlgorithmName hashAlgorithm in SupportedHashAlgorithms)
+            {
+                // hashAlgorithm, length, iterations, passwordHex, saltHex
+                yield return new object[] { hashAlgorithm.Name, 1, 1, s_passwordBytes.ByteArrayToHex(), s_salt.ByteArrayToHex() };
+                yield return new object[] { hashAlgorithm.Name, 1, 1, "", s_salt.ByteArrayToHex() };
+                yield return new object[] { hashAlgorithm.Name, 257, 257, "", s_salt.ByteArrayToHex() };
+                yield return new object[] { hashAlgorithm.Name, 257, 257, "D8D8D8D8D8D8D8D8", "D8D8D8D8D8D8D8D8" };
+                yield return new object[] { hashAlgorithm.Name, 257, 257, "0000000000000000", "0000000000000000" };
+            }
+
+            // Test around HMAC SHA1 and SHA256 block boundaries
```

Comment:
Hooray for edge case testing!

Reply:
Was there also a >2k-length password test, just to make sure that we don't simplify the code into producing an error?

Reply:
@bartonjs good idea, added a test with a large salt and large password.